File: Data Structures & Algorithms/combination-target-sum/submission-0.go
func combinationSum(nums []int, target int) [][]int {
	res := [][]int{}
	var dfs func(i int, total int, curr []int) 

	dfs = func(i int, total int, curr []int) {
		if total == target {
			fmt.Println(curr)
			snapshot := make([]int, len(curr))
			copy(snapshot, curr)
			res = append(res, snapshot)
			return
		}

		if i >= len(nums) || total > target {
			return
		}

		curr = append(curr, nums[i])
		dfs(i, total + nums[i], curr)

		curr = curr[:len(curr)-1]
		dfs(i+1, total, curr)
	}
	dfs(0, 0, []int{})
	return res
}

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []

        def dfs(candidates, curr):
            if not candidates: # length of candidates is 0
                res.append(curr)
                return

            
            for i in range(len(candidates)):
                dfs(candidates[:i] + candidates[i+1:],curr + [candidates[i]])

        dfs(nums, [])
        return res

            
        